### PR TITLE
[WIP] mfem CMake package

### PR DESCRIPTION
**Pull request by adrienbernede**
xSDK folks (represented by @acfisher) are having issues with the mfem spack package which relies on mfem makefile build system.
The idea was then to give a try at writing a Spack package for mfem that would use the CMake build system.

I did it the "Uberenv way", with a package that generates a configuration file.

This package seems in good shape. It gave me positive results building mfem with some dependencies, although PETSc is still causing some weird issues (https://lc.llnl.gov/gitlab/radiuss/radiuss-stack/-/pipelines/38216). However, I voluntarily adopted a step by step approach, and some things needs "closure".

My current attempt is to build Serac with mfem-cmake. This introduces Axom in the loop, and I'm having troubles building Axom with mfem-cmake, leading to errors like:

```
        950     /g/g91/bernede1/Projects/serac/uberenv_libs/gcc-8.3.1/netcdf-c-4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5internal.o): In function `nc4_reform_coord_var':
  >> 951     hdf5internal.c:(.text+0x9ef): undefined reference to `H5DSdetach_scale'
```

My initial guess was that I needed to add hdf5 as a required package for NetCDF in mfem. This was not done initially in MFEM CMake build system, but was mentioned:

```
 set(NETCDF_DIR "" CACHE PATH "Path to the NetCDF library.")
  8 # May need to add "HDF5" as requirement.
  7 set(NetCDF_REQUIRED_PACKAGES "" CACHE STRING
  6     "Additional packages required by NetCDF.")
```

I also had voluntarily skipped all the "required packages" in the configuration file generated by spack as part of the step-by-step approach.

This made me realize that Axom `FindMFEM.cmake` is not looking for the MFEMConfig.cmake, but for its makefile equivalent. So I patched the axom version used by Serac to allow this. But Axom still fails to find it, although it's there.

### Conclusion

This is a call for help on this, as I need to know if this CMake approach is working, and Serac has been identified as a good test candidate. I'm especially thinking of @white238 and @joshessman-llnl for this.

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
Note: To reproduce the problem on Gitlab, trigger a pipeline for that branch with the variable `BUILD_DEPS` set to `ON`.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -7,8 +7,8 @@ variables:
   GIT_SUBMODULE_STRATEGY: recursive
   PROJECT_ALLOC_NAME: ${CI_PROJECT_NAME}_ci_${CI_PIPELINE_ID}
   BUILD_ROOT: ${CI_PROJECT_DIR}
-  SLURM_ACCOUNT: -p pdebug -A eng
-  LSF_ACCOUNT: -q pdebug -G eng
+  SLURM_ACCOUNT: -p pdebug -A radiuss
+  LSF_ACCOUNT: -q pdebug -G radiuss
```
Comment (white238) [inline .gitlab-ci.yml]: Is there a way to figure out who can charge this account? This could solve Axom's problem as well

Comment (adrienbernede) [reply .gitlab-ci.yml]: Sorry, that's a left-over of me not being able to use "eng".

Comment (white238) [reply .gitlab-ci.yml]: got me all excited....

Comment (adrienbernede) [reply .gitlab-ci.yml]: I can tell :)

File: scripts/spack/packages/mfem-cmake/package.py [hunk 2]
```diff
@@ -0,0 +1,895 @@
...
+            gslib_dir = get_spec_path(spec, "gslib")
+            cfg.write(cmake_cache_entry("GSLIB_DIR", gslib_dir))
+
+        if '+netcdf' in spec:
+            netcdf_dir = get_spec_path(spec, "netcdf-c")
+            cfg.write(cmake_cache_entry("NETCDF_DIR", netcdf_dir))
+
+            #TODO (bernede1@llnl.gov): what about NETCDF_REQUIRED_PACKAGES
+            # see MFEM config/defaults.cmake
+            cfg.white(cmake_cache_string("NetCDF_REQUIRED_PACKAGES", "hdf5"))
```
Comment (joshessman-llnl) [inline scripts/spack/packages/mfem-cmake/package.py]:
```suggestion
            cfg.write(cmake_cache_string("NetCDF_REQUIRED_PACKAGES", "hdf5"))
```

**Comment by adrienbernede** [comment]
Thanks @joshessman-llnl ... now I'm embarrassed ;)
Also found that I need to uninstall mfem for the changes in the package to take effect. Will let you know.

**Comment by joshessman-llnl** [comment]
> 
> 
> Thanks @joshessman-llnl ... now I'm embarrassed ;)
> Also found that I need to uninstall mfem for the changes in the package to take effect. Will let you know.

I tried this on LC and observed the following:
```
-- Looking for NetCDF ...
--    in NETCDF_DIR = /usr/WS2/smithdev/libs/serac/toss_3_x86_64_ib/2021_03_24_12_12_48/gcc-8.3.1/netcdf-c-4.7.4
-- NetCDF: looking for required package: hdf5
CMake Error at config/cmake/modules/MfemCmakeUtilities.cmake:429 (find_package):
  By not providing "Findhdf5.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "hdf5", but
  CMake did not find one.

  Could not find a package configuration file provided by "hdf5" with any of
  the following names:

    hdf5Config.cmake
    hdf5-config.cmake

  Add the installation prefix of "hdf5" to CMAKE_PREFIX_PATH or set
  "hdf5_DIR" to a directory containing one of the above files.  If "hdf5"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  config/cmake/modules/FindNetCDF.cmake:18 (mfem_find_package)
  CMakeLists.txt:288 (find_package)
```

I'm going to try changing `NetCDF_REQUIRED_PACKAGES` from `hdf5` to `HDF5` as from my understanding of how the MFEM macros work the latter will use the more specific https://cmake.org/cmake/help/latest/module/FindHDF5.html

**Comment by adrienbernede** [comment]
I had the same, and currently trying to set HDF5_DIR using the spec in spack.

**Comment by adrienbernede** [comment]
Putting the cart before the horses here. So applying the same change as you, and imposing hdf5 location.

**Comment by joshessman-llnl** [comment]
> 
> 
> Putting the cart before the horses here. So applying the same change as you, and imposing hdf5 location.

I think the FindHDF5 module uses HDF5_ROOT and not HDF5_DIR:
https://github.com/Kitware/CMake/blob/88060f471766222c54a080900f73abf29586a6f4/Modules/FindHDF5.cmake#L491-L498

though there's some CMake version wrangling that needs to happen as well to take advantage of this:
https://github.com/LLNL/axom/blob/160748c6115b0ec712fd06ad30e4d265d0294453/src/cmake/thirdparty/SetupHDF5.cmake#L20-L27

Spack *typically* will set up the environment such that these things aren't necessary but that may not be guaranteed.

**Comment by adrienbernede** [comment]
> FindHDF5 module uses HDF5_ROOT and not HDF5_DIR.

Good point, tried both but I'm stopped by a missing compiler for C. Trying to activate C in MFEM CMake system. 

> though there's some CMake version wrangling

True, I'm guessing Axom gives the priority to the location defined in their configuration file, which is typically set by their Spack package based on the location on HDF5 in spack context.
https://github.com/spack/spack/blob/6997991ad2ba58f2c70515a36e985a7c70fc1b2f/var/spack/repos/builtin/packages/axom/package.py#L283

> Spack typically will set up the environment such that these things aren't necessary but that may not be guaranteed.

I'm trying to stay away from assumptions on spack behavior: once this is working, I could consider "optimizations" :)

**Comment by acfisher** [comment]
It looks like Adrien needs to pivot to another project for now.  Can the Serac+Axom folks handle this integration or does something need to change in the MFEM spack package?

**Comment by adrienbernede** [comment]
My latest findings:
- XSDK_ENABLE_C in MFEM CMake wasn't working for me. I had to enforce C as a language next to CXX.
- I had to apply the new behavior for policies 0053 and 0054.

But I have to admit I rushed those and did not dig much deep.

In the end, I still have undefined referenced to hdf5 symbols through netcdf when building Axom. I am guessing MFEM CMake build system needs to be updated to export the appropriate flags to link, but again, running out of time (and skill) on this task, at least for a little while.

**Comment by adrienbernede** [comment]
Shortcut: we need the hdf5 link args in `MFEM_EXT_LIBS` in the config.mk.

**Comment by joshessman-llnl** [comment]
> 
> 
> Shortcut: we need the hdf5 link args in `MFEM_EXT_LIBS` in the config.mk.

It looks like the "base" hdf5 library is pulled in fine but the high-level (hdf5_hl.a) library is not.   The GNU Make build system takes this into account:

https://github.com/mfem/mfem/blob/b467c70222aa965d906dc8e2e6eaab236f799934/config/defaults.mk#L304-L306

I've put together a few patches in a branch on top of this one: https://github.com/LLNL/serac/tree/testing/essman/mfem_cmake
and I was able to get serac to build on quartz.  I anticipate that the lassen build will require further patching but that's my next objective.

**Comment by adrienbernede** [comment]
I noticed the difference between the management of NetCDF dependency in makefile and cmake build systems. But I wasn't sure how to handle this.
I think I understand what you've done. The gray zone being why you needed to temporarily erase CMAKE_MODULE_PATH: mfem does not have a FindHDF5.cmake. Does it means CMake wouldn't look into it's own location if the module is not found in the specified path?

**Comment by joshessman-llnl** [comment]
> The gray zone being why you needed to temporarily erase CMAKE_MODULE_PATH: mfem does not have a FindHDF5.cmake. Does it means CMake wouldn't look into it's own location if the module is not found in the specified path?

I needed to "patch" the HDF5_LIBRARIES to add `hdf5_hl`, and the mfem_find_package function doesn't provide any way of doing this (the fact that it's a function and not a macro also makes the patching more challenging).  I chose to extend the "find" logic for hdf5 by adding a `FindHDF5.cmake` to MFEM (that's the file I added the CMAKE_MODULE_PATH stuff to).  The extra logic I need is just a small addition on top of CMake's built-in FindHDF5, so I call that one (erasing the MODULE_PATH to avoid infinite recursion due to the lower priority of the built-in) and then patch accordingly.

**Comment by jamiebramwell** [comment]
I'm worried we will be tasked with maintaining this forked spack package (and possibly the mfem cmake build). Should this be opened as a PR in the main spack repo as a WIP? @acfisher @adrienbernede

**Comment by adrienbernede** [comment]
@jamiebramwell I understand your concern ;)

I only created this branch to ease the testing with Serac. Uberenv is a great way to experiment with changes in Spack in a controlled environment. But the "real" PR in Spack is there: https://github.com/spack/spack/pull/22521

So, the reason for this PR is only because Serac (and incidentally Axom) are the first candidates for testing this. The goal is then to have the xSDK use it for their bundle package.

Maybe @acfisher can also try to reassure you ;).

**Comment by jamiebramwell** [comment]
@adrienbernede Ahhh that makes sense! I didn't realize a corresponding spack PR existed. As long as we follow up on that and ensure it eventually gets merged, I agree we're probably a good test particle. Thanks!

**Comment by adrienbernede** [comment]
Added a "do-not-merge" label to signify this is not meant to be merge in Serac.

**Pull request closed without merging**